File: backend/src/main/java/com/commerceops/erp/domain/notification/service/NotificationService.java
package com.commerceops.erp.domain.notification.service;

import com.commerceops.erp.domain.notification.dto.NotificationResponse;
import com.commerceops.erp.domain.notification.dto.UnreadNotificationCountResponse;
import com.commerceops.erp.domain.notification.entity.Notification;
import com.commerceops.erp.domain.notification.enums.NotificationType;
import com.commerceops.erp.domain.notification.repository.NotificationRepository;
import com.commerceops.erp.domain.user.entity.User;
import com.commerceops.erp.global.exception.BusinessException;
import com.commerceops.erp.global.exception.ErrorCode;
import com.commerceops.erp.global.response.PageResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class NotificationService {

    private final NotificationRepository notificationRepository;

    @Transactional
    public void notifyUser(User user, NotificationType type, String title, String message,
                           String targetType, Long targetId) {
        notificationRepository.save(Notification.builder()
                .user(user)
                .type(type)
                .title(title)
                .message(message)
                .targetType(targetType)
                .targetId(targetId)
                .build());
    }

    public PageResponse<NotificationResponse> getMyNotifications(User user, int page, int size) {
        return PageResponse.from(notificationRepository
                .findByUserOrderByCreatedAtDesc(user, PageRequest.of(page, size, Sort.by("createdAt").descending()))
                .map(NotificationResponse::from));
    }

    public PageResponse<NotificationResponse> getAdminNotifications(int page, int size, boolean unreadOnly) {
        PageRequest pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        return PageResponse.from((unreadOnly
                ? notificationRepository.findByReadAtIsNull(pageable)
                : notificationRepository.findAll(pageable))
                .map(NotificationResponse::from));
    }

    public UnreadNotificationCountResponse getUnreadCount(User user) {
        return new UnreadNotificationCountResponse(notificationRepository.countByUserAndReadAtIsNull(user));
    }

    @Transactional
    public NotificationResponse markRead(User user, Long notificationId) {
        Notification notification = notificationRepository.findById(notificationId)
                .orElseThrow(() -> new BusinessException(ErrorCode.INVALID_INPUT_VALUE));
        if (!notification.getUser().getId().equals(user.getId())) {
            throw new BusinessException(ErrorCode.FORBIDDEN);
        }
        notification.markRead();
        return NotificationResponse.from(notification);
    }

    @Transactional
    public UnreadNotificationCountResponse markAllRead(User user) {
        notificationRepository.markAllReadByUser(user);
        return getUnreadCount(user);
    }
}

File: frontend/src/lib/services/notificationService.ts
import { apiClient, PageResponse } from '@/lib/api';

export interface ApiNotification {
  id: number;
  userId: number;
  type: 'ORDER_STATUS' | 'INQUIRY_ANSWERED' | 'RETURN_PROCESSED' | 'SYSTEM';
  title: string;
  message: string;
  targetType: string | null;
  targetId: number | null;
  read: boolean;
  readAt: string | null;
  createdAt: string;
}

export interface ApiUnreadNotificationCount {
  unreadCount: number;
}

export const notificationService = {
  getMyNotifications: (page = 0, size = 20) =>
    apiClient<PageResponse<ApiNotification>>(`/notifications?page=${page}&size=${size}`),

  getUnreadCount: () =>
    apiClient<ApiUnreadNotificationCount>('/notifications/unread-count'),

  markRead: (notificationId: number) =>
    apiClient<ApiNotification>(`/notifications/${notificationId}/read`, { method: 'PATCH' }),

  markAllRead: () =>
    apiClient<ApiUnreadNotificationCount>('/notifications/read-all', { method: 'PATCH' }),

  getAdminNotifications: (page = 0, size = 20, unreadOnly = false) =>
    apiClient<PageResponse<ApiNotification>>(
      `/admin/notifications?page=${page}&size=${size}&unreadOnly=${unreadOnly}`
    ),
};
